userspace/libsinsp/test/filter_parser.ut.cpp

```diff
@@ -0,0 +1,158 @@
+	test_accept("(a.a exists and b)");
+	test_accept("(a.a exists) and (b)");
+	test_accept("a.a exists and b");
+	test_accept("a.a=1 or b.b=2 and c");
+	test_accept("not (a)");
+	test_accept("not (not (a))");
+	test_accept("not (a.b=1)");
+	test_accept("not (a.a exists)");
+	test_accept("not a");
+	test_accept("a.b = 1 and not a");
+	test_accept("not not a");
+	test_accept("(not not a)");
+	test_accept("not a.b=1");
+	test_accept("not a.a exists");
+	test_accept("a.b = bla");
+	test_accept("a.b = 'bla'");
+	test_accept("a.b = not");
+	test_accept("a.b contains bla");
+	test_accept("a.b icontains 'bla'");
+	test_accept("a.g in (1, 'a', b)");
+	test_accept("evt.dir=> and fd.name=*.log");
+	test_accept("a.g in (1, 'a', b.c)");
+	test_accept("a.b = a.a");
+
+	// marked as bad in Falco smoke checks, but they should be good instead
+	test_accept("evt.arg[0] contains /bin");
+	test_accept("evt.arg[a] contains /bin");
+
+	// bad
+	test_reject("evt.arg[] contains /bin");
+	test_reject("a.b = b = 1");
+	test_reject("(a.b = 1");
+	test_reject("a.a invalidoperator xxx");
+
+	// marked as good in Falco smoke checks, but they should be bad instead
+	test_reject("a.g in ( 1 ,, , b)");
+	test_reject("#a and b; a and b");
+	test_reject("#a and b; # ; ; a and b");
+	test_reject("evt.dir=> and fd.name=/var/lo);g/httpd.log");
+	test_reject("notz and a and b");
+}
+
+// complex test case with all supported node types
+TEST(parser, expr_all_node_types)
+{
+	test_equal_ast(
+		"evt.name exists and evt.type in (a, b) and not evt.dir=< or proc.name=cat",
+		new or_expr({
+			new and_expr({
+				new unary_check_expr("evt.name", "", "exists"), 
+				new binary_check_expr("evt.type", "", "in", new list_expr({"a", "b"})),
+				new not_expr(
+					new binary_check_expr("evt.dir", "", "=", new value_expr("<"))
+				),
+			}),
+			new binary_check_expr("proc.name", "", "=", new value_expr("cat")),
+		})
+	);
+}
+
+// complex example with parenthesis
+TEST(parser, expr_parenthesis)
+{
+	test_equal_ast(
+		"evt.name exists and evt.type in (a, b) and not evt.dir=< or proc.name=cat",
+		new or_expr({
+			new and_expr({
+				new unary_check_expr("evt.name", "", "exists"), 
+				new binary_check_expr("evt.type", "", "in", new list_expr({"a", "b"})),
+				new not_expr(
+					new binary_check_expr("evt.dir", "", "=", new value_expr("<"))
+				),
+			}),
+			new binary_check_expr("proc.name", "", "=", new value_expr("cat")),
+		})
+	);
+}
+
+// stressing nested negation and identifiers
+TEST(parser, expr_multi_negation)
+{
+	test_equal_ast(
+		"evt.name exists and evt.type in (a, b) and not evt.dir=< or proc.name=cat",
+		new or_expr({
+			new and_expr({
+				new unary_check_expr("evt.name", "", "exists"), 
+				new binary_check_expr("evt.type", "", "in", new list_expr({"a", "b"})),
+				new not_expr(
+					new binary_check_expr("evt.dir", "", "=", new value_expr("<"))
+				),
+			}),
+			new binary_check_expr("proc.name", "", "=", new value_expr("cat")),
+		})
+	);
+}
```

Comment:
That's a good point. I investigated the anatomy of Falco and sinsp ASTs extensively during my research, before coming up with this grammar. The only perceivable difference is that Falco's AST produced with Lua `lpeg` are always binary (e.g. each `and`/`or` node has always two children), whereas in this case the boolean-op nodes can have more than two children. This design choice allows to reduce the recursion depth required to traverse ASTs, and makes it easier to visit them and eventually compact/simplify them with some optimization in the future.

What really matters, in this case, is that the check order is preserved, and I made sure it was with the new AST nodes. Will try to bring more data to support this point.

Reply:
I'll bring an example from an issue opened a long ago (see https://github.com/falcosecurity/falco/issues/706). Consider the filter of macro `open_write` taken from the default Falco ruleset:
```
(evt.type=open or evt.type=openat) and evt.is_open_write=true and fd.typechar='f' and fd.num>=0
```

The AST dump presented in https://github.com/falcosecurity/falco/issues/706#issuecomment-508802953 shows how Falco processes the filter of `open_write`. On the left, I report the AST reported in that response, on the right I report the result of the newly proposed parser (trying to reproduce a similar syntax. The printing code is just a draft and is not pushed yet):

<table>
<tr>
<th>
Falco Parser AST
</th>
<th>
New Parser AST
</th>
</tr>

<tr>

<td>
<pre>
{
  open_write = {
    ast = {
      left = {
        left = {
          left = {
            left = {
              left = {
                type = "FieldName",
                value = "evt.type"
              },
              operator = "=",
              right = {
                type = "BareString",
                value = "open"
              },
              type = "BinaryRelOp"
            },
            operator = "or",
            right = {
              left = {
                type = "FieldName",
                value = "evt.type"
              },
              operator = "=",
              right = {
                type = "BareString",
                value = "openat"
              },
              type = "BinaryRelOp"
            },
            type = "BinaryBoolOp"
          },
          operator = "and",
          right = {
            left = {
              type = "FieldName",
              value = "evt.is_open_write"
            },
            operator = "=",
            right = {
              type = "BareString",
              value = "true"
            },
            type = "BinaryRelOp"
          },
          type = "BinaryBoolOp"
        },
        operator = "and",
        right = {
          left = {
            type = "FieldName",
            value = "fd.typechar"
          },
          operator = "=",
          right = {
            type = "String",
            value = "f"
          },
          type = "BinaryRelOp"
        },
        type = "BinaryBoolOp"
      },
      operator = "and",
      right = {
        left = {
          type = "FieldName",
          value = "fd.num"
        },
        operator = ">=",
        right = {
          type = "Number",
          value = 0
        },
        type = "BinaryRelOp"
      },
      type = "BinaryBoolOp"
    },
    used = false
  }
}
</pre>
</td>
<td>
<pre>
{
  bool_op = "and"
  children = [
    {
      bool_op = "or"
      children = [
        {
          field = "evt.type"
          field_arg = ""
          op = "="
          value = "open"
        },
        {
          field = "evt.type"
          field_arg = ""
          op = "="
          value = "openat"
        },
      ],
    },
    {
      field = "evt.is_open_write"
      field_arg = ""
      op = "="
      value = "true"
    },
    {
      field = "fd.typechar"
      field_arg = ""
      op = "="
      value = "f"
    },
    {
      field = "fd.num"
      field_arg = ""
      op = ">="
      value = "0"
    },
  ],
}
</pre>
</td>
</tr>
</table>

From this, I think we can conclude that:
1. Falco's parser creates binary left-associative AST nodes for boolean operators, whereas the new parser creates non-binary non-associative nodes. This has the benefit of creating more compact ASTs, and reduces the average recursion depth.
2. Considering a left-most depth-first traversal, the two ASTs produce equivalent boolean formulas.

Reply:
> I think this is a big enough change that we might want to have it initially available via a feature flag. I'm just worried that we end up with some strange corner case wrt condition parsing where this implementation ends up being different, and there's no quick way to revert to the "old" behavior.

Hey @mstemm 
I have a solid opinion not to use a future flag because the burden of maintaining both implementations (even just for a couple of months) is too much. Moreover, fixing a bug in the new implementation is way simpler and faster than touching the old implementation (which honestly is known to be buggy and very difficult to modify).

I instead spend more time testing before merging this PR, although the test @jasondellaluce  did in Falco seems to be already enough.

Reply:
Thanks, that summary helps a lot! I agree that the new AST (especially chaining logical operators using a list) looks more compact. Also the additional tests help show good coverage.